src/aggregator/aggregator/leader_flush_mgr.go

```diff
@@ -395,12 +403,49 @@ func cloneForwardedFlushTimesForResolution(
 }
 
 type leaderFlushTask struct {
-	mgr      *leaderFlushManager
-	duration tally.Timer
-	flushers []flushingMetricList
+	sync.RWMutex
+
+	wg   sync.WaitGroup
+	done bool
+
+	mgr               *leaderFlushManager
+	flushTimesManager FlushTimesManager
+	duration          tally.Timer
+	flushers          []flushingMetricList
+	flushTimes        *schema.ShardSetFlushTimes
+}
+
+func (t *leaderFlushTask) Complete() bool {
+	t.RLock()
+	defer t.RUnlock()
+	return t.done
+}
+
+func (t *leaderFlushTask) WaitComplete() {
+	t.wg.Wait()
+}
+
+func (t *leaderFlushTask) runComplete(took time.Duration) {
+	if t.flushTimes != nil {
+		if err := t.flushTimesManager.StoreAsync(t.flushTimes); err != nil {
+			t.mgr.logger.Error("could not store flush times", zap.Error(err))
+		}
+	}
+
+	t.duration.Record(took)
+
+	t.Lock()
+	t.done = true
+	t.Unlock()
+
+	// Record task done.
+	t.wg.Done()
```

Comment:
Would it make more sense to record done first to unblock, then set done? Should be fine in practice probably though